Add cg_damagePlums

ernie opened the pull request:
Shows a quake live style damage plum on damage when set to 1.

ernie committed:
Add cg_damagePlums

Shows a quake live style damage plum on damage when set to 1.

WofWca commented on code/game/g_combat.c:
```diff
@@ -22,6 +22,37 @@ void ScorePlum( gentity_t *ent, vec3_t origin, int score ) {
+	plum->r.singleClient = attacker->s.number;
+	//
+	plum->s.otherEntityNum = attacker->s.number;
+	plum->s.time = damage;
```
I wonder if we really need a server event for this? Can't the client do this not far from where the "hit" sound is played, i.e. `CG_TransitionPlayerState`?

https://github.com/ec-/baseq3a/blob/aa8dac0e13b2c149b6d196882482e8e1429bc8ab/code/cgame/cg_playerstate.c#L483-L527

WofWca reviewed (commented):
I'm a little concerned with performance, but otherwise looks sensible in my amateur eyes.
But the performance, besides RAM usage, is not affected when the option is disabled, and is disabled by default, so it's not a big blocker?
I didn't test this yet though.

The-Gig commented:
Just wondering: could enabling this be considered an unfair advantage over who plays with default settings? _Just wondering_ whether it might be the case to allow server admins to disallow its usage (or even to force it to everyone?), via g_ cvar or dmflag?

ernie commented on code/game/g_combat.c:
```diff
@@ -22,6 +22,37 @@ void ScorePlum( gentity_t *ent, vec3_t origin, int score ) {
+
+	// check if the client supports damage plums
+	trap_GetUserinfo( attacker->s.number, userinfo, sizeof( userinfo ) );
+	damagePlumsEnabled = atoi( Info_ValueForKey( userinfo, "cg_damagePlums" ) );
```
I'm flexible on approach here if we decide we'd like to implement the feature. I did the work for another repo and wanted to see if there was interest here.

ernie committed:
Only parse cg_damagePlums userinfo on change

ernie commented:
@WofWca took your suggestion. wdyt?

As for the idea of disallowing it, I'm not sure the damage numbers are likely to be a competitive advantage much beyond the damage sounds, which aren't server-toggled?

WofWca commented on code/cgame/cg_localents.c:
```diff
@@ -870,6 +870,110 @@ void CG_AddScorePlum( localEntity_t *le ) {
+	// if the view would be "inside" the sprite, kill the sprite
+	VectorSubtract( origin, cg.refdef.vieworg, delta );
+	len = VectorLength( delta );
+	if ( len < 20 ) {
```
```suggestion
	len = VectorLengthSquared( delta );
	if ( len < 20*20 ) {
```

`VectorLengthSquared` is cheaper, is there a reason you changed this?

WofWca commented on code/cgame/cg_effects.c:
```diff
@@ -407,10 +407,10 @@ void CG_ScorePlum( int client, const vec3_t origin, int score ) {
 	le->lifeRate = 1.0 / ( le->endTime - le->startTime );
 
-	
+
```
Let's not include unrelated formatting changes?


```suggestion
	
```

WofWca reviewed (approved):
To my untrained eye of a stranger on the internet, looks good!

I have tested the feature for a while. Some feedback:
- In Quake Live, the size of the numbers is always the same regardless of distance.
  I think this is nice. Can we achieve the same?
  Does the game even support this?
- At close distance it's hard to see the number.
  I think the origin should be moved a little down,
  closer to the center of the player.
  Or maybe also the "raising" of the number should be tuned down?
- Maybe the font needs to be made smaller?
- The aggregation seems like a bad thing sometimes?
  e.g. when shooting with the lightning gun, you sometimes get a plum of 16
  instead of 2 plums of 8.
  This throws you off a little.

But overall I think this is about good to merge.
Just take a look at whether you'd want to address my smaller comments.
The rest can be addressed in follow-up MRs.

Edit: for those lazy, here is the build of the mod with this MR's changes (just rename to `pak8a.pk3`): 
[pak8a.pk3](https://github.com/user-attachments/files/24532090/pak8a.pk3.zip)


https://github.com/user-attachments/assets/90f35e95-b682-49bc-8ed0-26e1089063d2

https://github.com/user-attachments/assets/d517f498-83ca-46ad-a334-a05ea4d37fed

ernie committed:
Follow Quake Live thresholds

Also, adjust to match colors and set a consistent size. Spawn entity at
about half height. Overall, should look closer to Quake Live appearance.

WofWca commented:
Yes, I think this looks better! For comparison, Quake Live (don't look at the bottom right corner though).

![converted-1](https://github.com/user-attachments/assets/93089352-9434-4146-95a5-9addd9f063c2) ![converted-6](https://github.com/user-attachments/assets/a01fa113-e7e0-4a9b-9408-a8c85246195a)

Indeed it looks like that the font family is the reason that the numbers are more readable in Quake Live.

And here is a build: [pak8a.pk3](https://github.com/user-attachments/files/24549468/pak8a.pk3.zip).

<details><summary>More screenshots</summary>

![converted-2](https://github.com/user-attachments/assets/f86dc2ec-9609-4b9c-84a7-2e1dbf376ae4)
![converted-10](https://github.com/user-attachments/assets/0a1e3daf-6040-41ef-9c32-419b322ac023)


</details>

WofWca commented on code/game/g_combat.c:
```diff
@@ -22,6 +22,29 @@ void ScorePlum( gentity_t *ent, vec3_t origin, int score ) {
+void DamagePlum( gentity_t *attacker, vec3_t origin, int damage ) {
+	gentity_t *plum;
+	char userinfo[MAX_INFO_STRING];
+	int damagePlumsEnabled;
```
```suggestion
```

Discovered by C compiler.

WofWca reviewed (changes requested):
Hey, I think I noticed something weird with this MR. If you go to single player, e.g. the first level (with Crash), the bots will take way longer (like ~15 seconds) to respawn.
This also happens with `cg_damagePlums 0`.
I can't reproduce this on the current main branch.

I'm not sure what could be the cause of this. Maybe the change to `EV_MAX`?

WofWca reviewed (approved):
OK, apologies for false accusations. I just reproduced it on the main branch as well. I think this could have to do with `sv_fps 60`.

ernie committed:
Update code/game/g_combat.c

Remove unused variables.

Co-authored-by: WofWca <wofwca@protonmail.com>

WofWca commented:
Another thing I noticed about this: old game can't play back demos recorded on the new game (this MR). It errors out with "unknown event" as soon as there is a damage plum in the demo playback. IDK if this is important.

WofWca commented on code/game/bg_events.h:
```diff
@@ -108,7 +108,9 @@ DECLARE_EVENT( EV_TAUNT_NO ),
-DECLARE_EVENT( EV_TAUNT_PATROL )
+DECLARE_EVENT( EV_TAUNT_PATROL ),
+
+DECLARE_EVENT( EV_DAMAGEPLUM )				// damage plum
```
Hey, I've come up with a way to keep network compatibility!
This will display damage plums as score plums in older clients.

<img width="400" height="714" alt="Image" src="https://github.com/user-attachments/assets/f838c2b1-5247-4f2c-a7d7-d4109b98736c" />

[0001-fix-don-t-introduce-new-event-reuse-EV_SCOREPLUM.patch](https://github.com/user-attachments/files/24993211/0001-fix-don-t-introduce-new-event-reuse-EV_SCOREPLUM.patch)

<details><summary>patch text</summary>

```patch
From 0cefcfcdc83db1ad2f9c263362ca704dca69cd40 Mon Sep 17 00:00:00 2001
From: WofWca <wofwca@protonmail.com>
Date: Sun, 1 Feb 2026 18:52:10 +0400
Subject: [PATCH] fix: don't introduce new event, reuse EV_SCOREPLUM

This fixes inability to play back demos that were recorded
in the new version on a client that doesn't know
about the `EV_DAMAGEPLUM` event. See
https://github.com/ec-/baseq3a/pull/56#issuecomment-3772877175.

This will display damage plums as score plums in older clients.
---
 code/cgame/cg_event.c | 10 +++++-----
 code/game/bg_events.h |  4 +---
 code/game/bg_public.h |  1 +
 code/game/g_combat.c  |  3 ++-
 4 files changed, 9 insertions(+), 9 deletions(-)

diff --git a/code/cgame/cg_event.c b/code/cgame/cg_event.c
index 15390dcb..fe33a047 100644
--- a/code/cgame/cg_event.c
+++ b/code/cgame/cg_event.c
@@ -960,11 +960,11 @@ void CG_EntityEvent( centity_t *cent, vec3_t position, int entityNum ) {
 #endif
 
 	case EV_SCOREPLUM:
-		CG_ScorePlum( cent->currentState.otherEntityNum, cent->lerpOrigin, cent->currentState.time );
-		break;
-
-	case EV_DAMAGEPLUM:
-		CG_DamagePlum( cent->lerpOrigin, cent->currentState.time );
+		if (cent->currentState.eventParm & SCOREPLUM_IS_DAMAGEPLUM) {
+			CG_DamagePlum( cent->lerpOrigin, cent->currentState.time );
+		} else {
+			CG_ScorePlum( cent->currentState.otherEntityNum, cent->lerpOrigin, cent->currentState.time );
+		}
 		break;
 
 	//
diff --git a/code/game/bg_events.h b/code/game/bg_events.h
index 27d74e09..b9e0b2ce 100644
--- a/code/game/bg_events.h
+++ b/code/game/bg_events.h
@@ -108,9 +108,7 @@ DECLARE_EVENT( EV_TAUNT_NO ),
 DECLARE_EVENT( EV_TAUNT_FOLLOWME ),
 DECLARE_EVENT( EV_TAUNT_GETFLAG ),
 DECLARE_EVENT( EV_TAUNT_GUARDBASE ),
-DECLARE_EVENT( EV_TAUNT_PATROL ),
-
-DECLARE_EVENT( EV_DAMAGEPLUM )				// damage plum
+DECLARE_EVENT( EV_TAUNT_PATROL )
 
 #ifdef EVENT_ENUMS
 	, DECLARE_EVENT( EV_MAX )
diff --git a/code/game/bg_public.h b/code/game/bg_public.h
index c5374159..af02efe6 100644
--- a/code/game/bg_public.h
+++ b/code/game/bg_public.h
@@ -355,6 +355,7 @@ typedef enum {
 #undef EVENT_ENUMS
 } entity_event_t;
 
+#define SCOREPLUM_IS_DAMAGEPLUM 0x01
 
 typedef enum {
 	GTS_RED_CAPTURE,
diff --git a/code/game/g_combat.c b/code/game/g_combat.c
index 4639789a..9156bc84 100644
--- a/code/game/g_combat.c
+++ b/code/game/g_combat.c
@@ -36,13 +36,14 @@ void DamagePlum( gentity_t *attacker, vec3_t origin, int damage ) {
 		return;
 	}
 
-	plum = G_TempEntity( origin, EV_DAMAGEPLUM );
+	plum = G_TempEntity( origin, EV_SCOREPLUM );
 	// only send this temp entity to the attacker
 	plum->r.svFlags |= SVF_SINGLECLIENT;
 	plum->r.singleClient = attacker->s.number;
 	//
 	plum->s.otherEntityNum = attacker->s.number;
 	plum->s.time = damage;
+	plum->s.eventParm = SCOREPLUM_IS_DAMAGEPLUM;
 }
 
 /*
-- 
2.41.0

```

Maybe one should consider hiding them completely, e.g. by providing an out-of-range `client` so that the `if (client != cg.predictedPlayerState.clientNum) return` check is triggered, but I think it's OK to show them together with regular score plums.
